File: CanteenServiceDatabaseServer/GrpcService/converters/MenuConverter.cs
using DTO;
using Google.Protobuf.Collections;
using grpcProtoFiles;

namespace GrpcService.converters; 

public class MenuConverter {

    public MenuDTO ToDto(Menu menu) {
        return new MenuDTO() {
            Id = menu.Id,
            Description = menu.Description,
            Ingredients = menu.Ingredients.ToList(),
            Name = menu.Name,
            Price = menu.Price
        };
    }

    public List<MenuDTO> ToDtoList(RepeatedField<Menu> menus) {
        return menus.Select(ToDto).ToList();
    }

    public Menu ToProto(MenuDTO menuDto) {
        Menu menu = new Menu() {
            Id = menuDto.Id,
            Description = menuDto.Description,
            Ingredients = {menuDto.Ingredients} ,
            Name = menuDto.Name,
            Price = menuDto.Price
        };

        return menu;    
    }

    public RepeatedField<Menu> ToProtoList(List<MenuDTO> menuDtos) {
        return new() {menuDtos.Select(ToProto).ToList()};
    }

}

File: CanteenService/src/main/java/com/example/canteenservice/grpc/converters/MenuConverter.java
package com.example.canteenservice.grpc.converters;

import com.example.canteenservice.dto.MenuDTO;
import grpcProtoFiles.MenuOuterClass;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class MenuConverter {

    public MenuDTO toDTO(MenuOuterClass.Menu menu){
        MenuDTO menuDTO = new MenuDTO();
        menuDTO.setId(menu.getId());
        menuDTO.setName(menu.getName());
        menuDTO.setDescription(menu.getDescription());
        menuDTO.setPrice(menu.getPrice());
        menuDTO.setIngredients(menu.getIngredientsList());
        return menuDTO;
    }

    public List<MenuDTO> toDTOList(List<MenuOuterClass.Menu> menus){
        // We map the list of menus to a list of menuDTOs
        return menus.stream().map(this::toDTO).toList();
    }

    public MenuOuterClass.Menu toProto(MenuDTO menuDTO){
        return MenuOuterClass.Menu.newBuilder()
                .setId(menuDTO.getId())
                .setName(menuDTO.getName())
                .setDescription(menuDTO.getDescription())
                .setPrice(menuDTO.getPrice())
                .addAllIngredients(menuDTO.getIngredients())
                .build();
    }
}
